### DISCUSSION TITLE
Creating Stripe "checkout_session" in application controller?

### DISCUSSION THREAD

**Discussion by dewey** [pay-rails/pay#710] [Q&A]
Hey, I'm looking to have various "upsell" modals across multiple views in my Rails project. I don't want to redirect to a "checkout" page, so in my case I'd need to have the checkout_session available in all controllers so I can use the session in my checkout button.

```
<%= render partial: "pay/stripe/checkout_button", locals: { session: @checkout_session, title: "Buy" } %>
```

I was told that I could put it in my application controller:

```
class ApplicationController < ActionController::Base
  before_action :set_payment_session

  def set_payment_session
    [ ... ]
  end
end
```

Later on I realized in the Stripe web interface that there's a lot of `billing_portal.session.created` web hook events which I assume are connected to having it in the ApplicationController. I'd assume that also has an influence on the page load. What would you recommend in that case?

Just link to a central checkout page or is there another way?

Thanks!

**Comment by excid3** [comment]
I would just have the modal submit to an action that creates a session and redirects to it. Then you have it nicely organized and you can still have the modals easily.

**Reply by dewey** [reply]
Hey Chris, thanks for the reply! I think I understand the approach. So I'd create a `checkouts` controller that could look like this:

```
class CheckoutsController < ApplicationController
  before_action :set_payment_session, :only => :create

  def set_payment_session
    if user_signed_in?
      current_user.set_payment_processor :stripe
      current_user.payment_processor.customer
    end
  end

  def create
    @checkout_session = current_user.payment_processor.checkout(
      mode: "subscription",
      line_items: "price_1Lmjb34987sRx0bAyt3gCKMk"
    )
    redirect_to @checkout_session.url, status: 303, allow_other_host: true
  end
end
```

In my views I'd  then just have calls like the following to open the Stripe Checkout. In the future if I'd have multiple products I'd just pass the line_items through the link_to and dynamically create the correct URLs.

```
    <%= link_to "buy", checkout_path, :action => 'create' %>
```

To confirm I'm understanding the usual approach: The `pay/stripe/checkout_button` partial is just responsible for dynamically creating the URL without a redirect by injecting the session-id into the `@checkout_session.url` directly, saving one round trip.

Thanks again for taking the time to respond to my initial question, I haven't even thought about this approach but it sounds like it will solve my issue.
